_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+
+In the simplest case, the app builds the full view tree, and that is diffed in full against the previous version. However, as UI scales, this would be inefficient, so there are *other* mechanisms to do finer grained change propagation, as described below.
+
+## Components
+
+The above is the basic architecture, enough to get started. Now we will go into some more advanced techniques.
+
+It would be very limiting to have a single "app state" type throughout the application, and require all callbacks to express their state mutations in terms of that global type. So we won't do that.
+
+The main tool for stitching together components is the `Adapt` view node. This node is so named because it adapts between one app state type and another, using a closure that takes mutable access to the parent state, and calls into a child (through a "thunk") with a mutable reference to the child state.
+
+In the simple case where the child component operates independently of the parent, the adapt node is a couple lines of code. It is also an attachment point for richer interactions - the closure can manipulate the parent state in any way it likes. The event handler of the child component is also allowed to return an arbitrary type (unit by default), for upward propagation of data.
+
+In Elm terminology, the Adapt node is similar to [Html map][Elm Html map], though it manipulates mutable references to state, as opposed to being a pure functional mapping between message types. It is also quite similar to the "lens" concept from the existing Druid architecture, and has some resemblance to [Binding][SwiftUI Binding] in SwiftUI as well.
+
+## Finer grained change propagation: memoizing
+
+Going back to the counter, every time the app logic is called, it allocates a string for the label, even if it's the same value as before. That's not too bad if it's the only thing going on, but as the UI scales it is potentially wasted work.
+
+Ron Minsky has [stated][Signals and Threads: Building a UI framework] "hidden inside of every UI framework is some kind of incrementalization framework." Xilem unapologetically contains at its core a lightweight change propagation engine, similar in scope to the attribute graph of SwiftUI, but highly specialized to the needs of UI, and in particular with a lightweight approach to *downward* propagation of dependencies, what in React would be stated as the flow of props into components.
+
+In this particular case, that incremental change propagation is best represented as a *memoization* node, yet another implementation of the View trait. A memoization node takes a data value (which supports both `Clone` and equality testing) and a closure which accepts that same data type. On rebuild, it compares the data value with the previous version, and only runs the closure if it has changed. The signature of this node is very similar to [Html.Lazy] in Elm.
```

Comment:
Previously you said that Xilem:
> removes some of the limitations. In particular, Druid requires app state to be clonable and diffable, a stumbling block for many new users

But this seems to be re-introducing that limitation if the user wants this finer grained change propagation.

Reply:
Yes. It's a tradeoff, and the difference is that it's now opt-in, while Druid absolutely required it.